docs: add dedicated router lifecycle and events guide (#62225)
angular/angular: adev/src/content/guide/routing/lifecycle-and-events.md

When you need to inspect a Router event sequence, you can enable logging for internal navigation events for debugging. You can configure this by passing a configuration option (`withDebugTracing()`) that enables detailed console logging of all routing events.

```ts
import { provideRouter, withDebugTracing } from '@angular/router';

const appRoutes: Routes = [];
bootstrapApplication(AppComponent,
  {
    providers: [
      provideRouter(appRoutes, withDebugTracing())
    ]
  }
);
```

For more information, check out the official docs on [`withDebugTracing`](api/router/withDebugTracing).

## Common use cases

Router events enable many practical features in real-world applications. Here are some common patterns that are used with router events.

### Loading indicators

Show loading indicators during navigation:

```typescript
import { Router, NavigationStart, NavigationEnd, NavigationCancel, NavigationError, NavigationSkipped } from '@angular/router';
import { inject } from '@angular/core';

export class AppComponent {
  private router = inject(Router);
  loading = false;

  constructor() {

>     this.router.events.subscribe(event => {

Reviewer: How confortable do we feel with sharing leaky examples like this one ? Should we include a `takeUntilDestroyed()` or would that add to much noise ?

Reviewer 2: I'd always recommend to include takeUntilDestroyed.

                 event instanceof NavigationError ||
                 event instanceof NavigationSkipped) {
        this.loading = false;
      }
    });
  }
}
```

### Analytics tracking

Track page views for analytics:

```typescript
import { Router, NavigationEnd } from '@angular/router';
import { inject } from '@angular/core';

export class AnalyticsComponent {
  private router = inject(Router);

  constructor() {
    this.router.events.subscribe(event => {
      if (event instanceof NavigationEnd) {
        // Send page view to analytics
        this.analytics.trackPageView(event.url);
      }
    });
  }
}
```

### Error handling

Handle navigation errors gracefully and provide user feedback:

```typescript
import { Component, inject } from '@angular/core';
import { Router, NavigationError, NavigationCancel } from '@angular/router';

@Component({
  selector: 'app-error-handler',
  standalone: true,
  template: `
    @if (errorMessage) {
      <div class="error-banner">
        {{ errorMessage }}
        <button (click)="dismissError()">Dismiss</button>
      </div>
    }
  `
})
export class ErrorHandlerComponent {
  private router = inject(Router);

>   errorMessage = '';

Author: Got it!

  constructor() {
    this.router.events.subscribe(event => {

>       if (event instanceof NavigationError) {

Reviewer 3: nit: should the error message be cleared on `NavigationStart`?

Author: Good point!

        console.error('Navigation error:', event.error);
        this.errorMessage = 'Failed to load page. Please try again.';
      } else if (event instanceof NavigationCancel) {
        console.warn('Navigation cancelled:', event.reason);

>         if (event.reason === 'GuardRejected') {

Author: I had no idea this existed. Thank you!

The following are common error events:

| Event                                             | Description                                                                      |
| ------------------------------------------------- | -------------------------------------------------------------------------------- |
| [`NavigationCancel`](api/router/NavigationCancel) | Occurs when the router cancels navigation. Often due to a guard returning false. |
| [`NavigationError`](api/router/NavigationError)   | Occurs when navigation fails. Could be due to invalid routes or resolver errors. |

For a list of all lifecycle events, check out the [complete table of this guide](#all-router-events).

## How to subscribe to router events

When you want to run code during specific navigation lifecycle events, you can do so by subscribing to the `router.events` and checking the instance of the event:

```ts
// Example of subscribing to router events
import { Component, inject, signal, effect } from '@angular/core';
import { Event, Router, NavigationStart, NavigationEnd } from '@angular/router';

> @Component()

Reviewer: nit: 
```suggestion
@Component({ ... })
```

## How to debug routing events

Debugging router navigation issues can be challenging without visibility into the event sequence. Angular provides a built-in debugging feature that logs all router events to the console, helping you understand the navigation flow and identify where issues occur.

When you need to inspect a Router event sequence, you can enable logging for internal navigation events for debugging. You can configure this by passing a configuration option (`withDebugTracing()`) that enables detailed console logging of all routing events.

```ts
import { provideRouter, withDebugTracing } from '@angular/router';

const appRoutes: Routes = [];
bootstrapApplication(AppComponent,
  {
    providers: [
      provideRouter(appRoutes, withDebugTracing())
    ]
  }
);
```

For more information, check out the official docs on [`withDebugTracing`](api/router/withDebugTracing).

## Common use cases

Router events enable many practical features in real-world applications. Here are some common patterns that are used with router events.

### Loading indicators

Show loading indicators during navigation:

> ```typescript

Reviewer: Because the ts include an inline template
```suggestion
```angular-ts
```

  template: `
    @if (loading()) {
      <div class="loading-spinner">Loading...</div>
    }
  `
})
export class AppComponent {
  private router = inject(Router);
  
  readonly loading = toSignal(
    this.router.events.pipe(
      map(() => !!this.router.getCurrentNavigation())
    ),
    { initialValue: false }
  );
}
```

### Analytics tracking

Track page views for analytics:

```typescript
import { Component, inject, signal, effect } from '@angular/core';
import { Router, NavigationEnd } from '@angular/router';

@Injectable({ providedIn: 'root' })
export class AnalyticsService {
  private router = inject(Router);
  private destroyRef = inject(DestroyRef);

  startTracking() {
    this.router.events.pipe(takeUntilDestroyed(this.destroyRef))
      .subscribe(event => {
        // Track page views when URL changes
        if (event instanceof NavigationEnd) {
           // Send page view to analytics
          this.analytics.trackPageView(url);
        }
      });
  }

  private analytics = {
    trackPageView: (url: string) => {
      console.log('Page view tracked:', url);
    }
  };
}
```

### Error handling

Handle navigation errors gracefully and provide user feedback:

> ```typescript

Reviewer: ```suggestion
```angular-ts
```

import { Component, inject, signal } from '@angular/core';
import { Router, NavigationStart, NavigationError, NavigationCancel, NavigationCancellationCode } from '@angular/router';

@Component({

>   selector: 'app-error-handler',
>   standalone: true,
>   template: `

Reviewer: ```suggestion
  selector: 'app-error-handler',
  template: `
```

    @if (errorMessage()) {
      <div class="error-banner">
        {{ errorMessage() }}
        <button (click)="dismissError()">Dismiss</button>
      </div>
    }
  `
})
export class ErrorHandlerComponent {
  private router = inject(Router);
  readonly errorMessage = signal('');

  constructor() {

>     this.router.events.subscribe(event => {

Reviewer: ```suggestion
    this.router.events.pipe(takeUntilDestroyed()).subscribe(event => {
```

The following are common error events:

| Event                                             | Description                                                                      |
| ------------------------------------------------- | -------------------------------------------------------------------------------- |
| [`NavigationCancel`](api/router/NavigationCancel) | Occurs when the router cancels navigation. Often due to a guard returning false. |
| [`NavigationError`](api/router/NavigationError)   | Occurs when navigation fails. Could be due to invalid routes or resolver errors. |

For a list of all lifecycle events, check out the [complete table of this guide](#all-router-events).

## How to subscribe to router events

When you want to run code during specific navigation lifecycle events, you can do so by subscribing to the `router.events` and checking the instance of the event:

```ts
// Example of subscribing to router events
import { Component, inject, signal, effect } from '@angular/core';
import { Event, Router, NavigationStart, NavigationEnd } from '@angular/router';

> @Component({
>   selector: 'app-router-events',
>   standalone: true,
>   template: ''
> })

Reviewer 2: ```suggestion
@Component()
```

export class RouterEventsComponent {
  private readonly router = inject(Router);
  

>   // Signal to track current navigation event
>   currentEvent = signal<Event | null>(null);
>   
>   constructor() {
>     // Subscribe to router events and update signal
>     this.router.events.subscribe((event: Event) => {
>       this.currentEvent.set(event);
>     });
>     
>     // React to event changes with effect
>     effect(() => {
>       const event = this.currentEvent();
>       if (event instanceof NavigationStart) {
>         // Navigation starting
>         console.log('Navigation starting:', event.url);
>       }
>       if (event instanceof NavigationEnd) {
>         // Navigation completed
>         console.log('Navigation completed:', event.url);
>       }
>     });
>   }

Reviewer 2: I dont think we need a signal and an effect to handle side effect on router events. Having it inside the subscribe callback should be good enough. But we need to have takeUntilDestroyed to cleanup the subscription.

```suggestion
  constructor() {
    // Subscribe to router events and react to events
    this.router.events.pipe(takeUntilDestroyed()).subscribe((event: Event) => {
      if (event instanceof NavigationStart) {
        // Navigation starting
        console.log('Navigation starting:', event.url);
      }
      if (event instanceof NavigationEnd) {
        // Navigation completed
        console.log('Navigation completed:', event.url);
      }
    });
  }
```

When you need to inspect a Router event sequence, you can enable logging for internal navigation events for debugging. You can configure this by passing a configuration option (`withDebugTracing()`) that enables detailed console logging of all routing events.

```ts
import { provideRouter, withDebugTracing } from '@angular/router';

const appRoutes: Routes = [];
bootstrapApplication(AppComponent,
  {
    providers: [
      provideRouter(appRoutes, withDebugTracing())
    ]
  }
);
```

For more information, check out the official docs on [`withDebugTracing`](api/router/withDebugTracing).

## Common use cases

Router events enable many practical features in real-world applications. Here are some common patterns that are used with router events.

### Loading indicators

Show loading indicators during navigation:

```typescript
import { Component, inject } from '@angular/core';
import { Router } from '@angular/router';
import { toSignal } from '@angular/core/rxjs-interop';
import { map } from 'rxjs/operators';

@Component({
  selector: 'app-loading',

>   standalone: true,

Reviewer 2: ```suggestion
```

  template: `
    @if (loading()) {
      <div class="loading-spinner">Loading...</div>
    }
  `
})
export class AppComponent {
  private router = inject(Router);
  

>   loading = toSignal(

Reviewer 2: ```suggestion
  readonly loading = toSignal(
```

    this.router.events.pipe(
      map(() => !!this.router.getCurrentNavigation())
    ),
    { initialValue: false }
  );
}
```

### Analytics tracking

Track page views for analytics:

```typescript
import { Component, inject, signal, effect } from '@angular/core';
import { Router, NavigationEnd } from '@angular/router';

> @Component({
>   selector: 'app-analytics',
>   standalone: true,
>   template: ''
> })
> export class AnalyticsComponent {

Reviewer 2: This probably would live in a service that is provided in root. And that service probably would have a `startTracking` method which would be called in AppComponent (that injects the service). 

```suggestion
@Injectable({ providedIn: 'root' })
export class AnalyticsService {
```

  private router = inject(Router);

>   private currentUrl = signal<string | null>(null);
> 
>   constructor() {
>     this.router.events.subscribe(event => {
>       if (event instanceof NavigationEnd) {
>         this.currentUrl.set(event.url);
>       }
>     });
> 
>     // Track page views when URL changes
>     effect(() => {
>       const url = this.currentUrl();
>       if (url) {
>         // Send page view to analytics
>         this.analytics.trackPageView(url);
>       }
>     });
>   }

Reviewer 2: Also as suggested above, we dont need to introduce a signal just for the sake of it. Instead we can do things inside the subscribe callback. 

```suggestion
  private destroyRef = inject(DestroyRef);

  startTracking() {
    this.router.events.pipe(takeUntilDestroyed(this.destroyRef))
      .subscribe(event => {
        // Track page views when URL changes
        if (event instanceof NavigationEnd) {
           // Send page view to analytics
          this.analytics.trackPageView(url);
        }
      });
  }
```

  private analytics = {
    trackPageView: (url: string) => {
      console.log('Page view tracked:', url);
    }
  };
}
```

### Error handling

Handle navigation errors gracefully and provide user feedback:

```typescript
import { Component, inject, signal } from '@angular/core';
import { Router, NavigationStart, NavigationError, NavigationCancel, NavigationCancellationCode } from '@angular/router';

@Component({
  selector: 'app-error-handler',
  standalone: true,
  template: `
    @if (errorMessage()) {
      <div class="error-banner">
        {{ errorMessage() }}
        <button (click)="dismissError()">Dismiss</button>
      </div>
    }
  `
})
export class ErrorHandlerComponent {
  private router = inject(Router);

>   errorMessage = signal('');

Reviewer 2: ```suggestion
  readonly errorMessage = signal('');
```

Reviewer 2: This would be the perfect case for `linkedSignal` to be honest. Since we want to derive the error message from the events, but also want to set it to empty string on user interaction. 

Should we go with it @atscott?
